File: Week5/BoilerCodes/ArrayVariation/stack.cpp
#include <cstdio>
#include <iostream>

class Stack {
    private:
        static const int MAX  {50};
        int items[MAX];
        int top;
    public:
        Stack();
        bool isEmpty();
        bool isFull();
        bool push(int item);
        int pop();
        int peek();
};

Stack::Stack() {
    top = -1;
}

bool Stack::isEmpty() {
    return top == -1;
}

bool Stack::isFull() {
    return top == MAX - 1;
}

bool Stack::push(int item) {
    if (isFull()) {
        printf("Stack Overflow! Cannot push %d\n", item);
        return false;
    }
    items[++top] = item;
    return true;
}

int Stack::pop() {
    if (isEmpty()) {
        std::cout << "Stack Underflow! Cannot pop from empty stack" << std::endl;
        return -1;
    }
    return items[top--];
}

int Stack::peek() {
    if (isEmpty()) {
        std::cout << "Stack is empty! Cannot peek" << std::endl;
        return -1;
    }
    return items[top];
}

File: Week5/BoilerCodes/LinkedListVariation/stack.go
package main

import (
	"errors"
	"fmt"
)

type Stack[T any] struct {
	items []T
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{
		items: make([]T, 0),
	}
}

func (s *Stack[T]) Push(item T) {
	s.items = append(s.items, item)
}

func (s *Stack[T]) Pop() (T, error) {
	var zero T
	if s.IsEmpty() {
		return zero, errors.New("stack is empty")
	}

	index := len(s.items) - 1
	item := s.items[index]
	s.items = s.items[:index]
	return item, nil
}

func (s *Stack[T]) Peek() (T, error) {
	var zero T
	if s.IsEmpty() {
		return zero, errors.New("stack is empty")
	}

	return s.items[len(s.items)-1], nil
}

func (s *Stack[T]) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack[T]) Size() int {
	return len(s.items)
}

func (s *Stack[T]) Clear() {
	s.items = make([]T, 0)
}

func (s *Stack[T]) String() string {
	return fmt.Sprintf("Stack{items: %v}", s.items)
}
